### ISSUE TITLE
Add guards for aliasing with side effects

### ISSUE THREAD

**Issue by jansel** [pytorch/pytorch#93722]
We currently handle aliasing of objects that (could) have side effects [here](https://github.com/facebookresearch/torchdynamo/blob/a9a894b8c514b1a0f239d4441fcb01e26f340870/torchdynamo/variables/builder.py#L85).  But there are no guards checking that the aliasing relationships stay the same in subsequent calls.  This could lead to incorrect behavior in an example like this:
```
def foo(left, right, x):
  left.append(x+1)
  right.append(x-1)
  return left, right

a, b, c = [], []. []
with torchdynamo.optimize(...):
   foo(a, b)  # this works fine

torchdynamo.reset()
with torchdynamo.optimize(...):
   foo(c, c)  # this works fine too

a, b, c = [], []. []
torchdynamo.reset()
with torchdynamo.optimize(...):
   foo(a, b)  # correct
   foo(c, c)  # aliasing relationship changed!  wrong behavior!

a, b, c = [], []. []
torchdynamo.reset()
with torchdynamo.optimize(...):
   foo(c, c)  # correct
   foo(a, b)  # aliasing relationship changed! wrong behaivor!
```

We should add a set of guards that check for this case and trigger a recompile.  We likely need some sort of hash table to do these checks quickly.  Adding similar protection for Tensor aliasing would likely also be useful for compiler backends, as it will allow them to generate code that relies on input aliasing being unchanged.

cc @ezyang @soumith @msaroufim @wconstab @ngimel @bdhirsh @voznesenskym @yanboliang @penguinwu @anijain2305 @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @desertfire

**Comment by jansel** [comment]
I believe so, though it is a pretty obscure one that is hard to hit.

**Comment by anijain2305** [comment]
@voznesenskym fixed this already with #104921
